File: p071m/simplify_path.js
/**
 * @param {string} path
 * @return {string}
 */
var simplifyPath = function (path) {
  const stack = [];
  for (let dir of path.split('/')) {
    if (dir === '' || dir === '.') {
      continue;
    } else if (dir === '..') {
      if (stack) stack.pop();
    } else {
      stack.push(dir);
    }
  }
  return '/' + stack.join('/');
};
// TEST
[
  ['', '/'],
  ['/', '/'],
  ['/a/b', '/a/b'],
  ['/home/', '/home'],
  ['/../', '/'],
  ['/home//foo/', '/home/foo'],
  ['/a/./b/../../c/', '/c'],
].forEach(([path, expected]) => {
  const actual = simplifyPath(path);
  console.log('Simplified canonical path of', path, '->', actual);
  console.assert(actual === expected);
});

File: p071m/simplify_path.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        stack = []
        for token in path.split("/"):
            if token == "." or token == "":
                continue
            elif token == "..":
                if stack:
                    stack.pop()
            else:
                stack.append(token)
        return "/" + "/".join(stack)


# TESTS
for path, expected in [
    ("", "/"),
    ("/", "/"),
    ("/a/b", "/a/b"),
    ("/home/", "/home"),
    ("/../", "/"),
    ("/home//foo/", "/home/foo"),
    ("/a/./b/../../c/", "/c"),
]:
    sol = Solution()
    actual = sol.simplifyPath(path)
    print("Simplified canonical path of", path, "->", actual)
    assert actual == expected
